### PR TITLE
reorganized tests and added missing one

### LINKED ISSUE #376: Missing tests
**Is your feature request related to a problem? Please describe.**
We have a few untested places.

**Describe the solution you'd like**
We have warnings about missing tests in the following places:
- [ ] `DateTimeService.Now`
- [ ] `HomeService.CheckTestStatus`
- [ ] `AccountController` *
- [x] `AdminController` *
- [x] `HomeController` *
- [x] `TestController` *
- [ ] `ManageController` *
- [ ] `Startup and StartupExtensions`  ??

*Those controllers will be covered probably in #357

### PR DESCRIPTION
**Pull request by szrtGreg**
## Description
Tests for Account Controller. There are missing 3 tests which will be done in other PR (see issue: #424 ). _"Done is better then perfect"_

## Where should the reviewer start?
AccoutControllerTests file

## Motivation and context
closes #376 
## How has this been tested?
All tests passed

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by szrtGreg** [comment]
@Mejtika Good Morning how would you write test for Login_ReturnsLoginViewModel ? I have trobules with mocking. Thanks!

obraz

obraz

**Comment by szrtGreg** [comment]
@Mejtika  You sent me suggestion (https://stackoverflow.com/questions/47198341/how-to-unit-test-httpcontext-signinasync?fbclid=IwAR2iU1Z4zHG6LViqlKiHVokY9yKaPCalrPnKTragWVTG3OkgL-y_I19BkbA) and it works! Thanks

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            // Act
+            var result = controller.Login(model, returnUrl);
+
+            // Assert
+            var allErrors = controller.ModelState.Values.SelectMany(v => v.Errors);
+            Assert.Single(allErrors);
+            Assert.Contains(allErrors, x => x.ErrorMessage == @EmailNotFound);
+            var viewResult = Assert.IsType<Task<IActionResult>>(result);
+
+            // Assert.IsType<LoginViewModel>(viewResult.Result)
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs:88]: please do not leave commented code

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        private static LoginViewModel GetLoginViewModel()
+        {
+            return new LoginViewModel
+            {
+                Email = "a@a.pl",
+                Password = "tajne",
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs:96]: please do not use polish words - use"password" or sth like that

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/HomeControllerTests.cs [hunk 3]
```diff
@@ -135,6 +136,23 @@ public void CheckTestStatus_ReturnsContentResultWithTestStatus()
...
+            var statusCode = 404;
+            var expectedMessage = ErrorMessages.NotFound;
+            using var controller = new HomeController(this._homeServiceMock.Object, this._adventService.Object);
+
+            // Act
+            var result = controller.Error(statusCode);
+
+            // Assert
+            var viewResult = Assert.IsType<ViewResult>(result);
+            var accualMessage = Assert.IsType<ErrorViewModel>(viewResult.ViewData.Model);
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/HomeControllerTests.cs:152]: typo, should be `actualMessage`

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/HomeControllerTests.cs [hunk 4]
```diff
@@ -135,6 +136,23 @@ public void CheckTestStatus_ReturnsContentResultWithTestStatus()
...
+            Assert.Equal(expectedMessage, accualMessage.Message.ToString());
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/HomeControllerTests.cs:153]: I suggest using `expectedMessage` and just `message`

**Comment by szrtGreg** [comment]
@WTobor Thank you for review. I did changes accordingly

**Comment by WTobor** [comment]
@szrtGreg is this PR ready? It is still a draft ;)

**Comment by szrtGreg** [comment]
Yes its still draft. I created PR in draft state only for core review. I had a lot of troubles with tests, thats why it takes me so long

**Comment by szrtGreg** [comment]
I have problem with Mocking SingInResult - Everytime it gets null value.
We have Account Controller  - Login Method

obraz

My test looks like that. And my problem with result varibale somehow I am still getting null value;
obraz

obraz


If anyone could help I would be appreciated

**Comment by WTobor** [comment]
@szrtGreg you are using _accountService_, not _accountServiceMock_

**Comment by WTobor** [comment]
@szrtGreg please resolve conflicts

**Comment by WTobor** [comment]
Maybe change
```
this._accountServiceMock.Setup(x => x.PasswordSignInAsync(model.Email, model.Password, model.RememberMe))
                                   .ReturnsAsync(SignInResult.Success);
```
to
```
this._accountServiceMock.Setup(x => x.PasswordSignInAsync(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<bool>()))
                                   .ReturnsAsync(SignInResult.Success);
```

**Comment by WTobor** [comment]
Aaaa! You are using in mock _model.Email_ - but in logic, we are using _user.UserName_

**Comment by WTobor** [comment]
@szrtGreg  remember to resolve conflicts :)

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs [hunk 5]
```diff
@@ -307,6 +307,93 @@ public void ExternalLoginCallback_ResustIsSuccessed_RedirectToLocal()
             Assert.Equal("Index", viewResult.ActionName);
         }
 
+        [Fact]
+        public void ExternalLoginCallback_ResustIsNotAllowed_RedirectToLocal()
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs:311]: what is Resust? :)

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,774 @@
...
+        {
+            // Arrange
+            ResetPasswordViewModel model = null;
+            using var controller = new AccountController(this._accountServiceMock.Object, this._loggerMock.Object);
+
+            // Act
+            Func<IActionResult> act = () => controller.ResetPassword(model).Result;
+
+            // Assert
+            Assert.Throws<AggregateException>(act);
```
Comment (Mejtika) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs:641]: The actual exception is ArgumentNullException. Using async/await avoids the AggregateException wrapper if the task has an exception. The code could look like this
`Func<Task<IActionResult>> act = async () => await controller.ResetPassword(model).ConfigureAwait(false); `
`await Assert.ThrowsAsync<ArgumentNullException>(act).ConfigureAwait(false);`

Comment (szrtGreg) [reply src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs:641]: He I did changes as you suggested

obraz

but there is stm wrong as test passed with different kind of errors eg:

obraz

etc.

Comment (Mejtika) [reply src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs:641]: In this case you need to also await that assertion

Comment (szrtGreg) [reply src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs:641]: Thanks!

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs [hunk 7]
```diff
@@ -0,0 +1,743 @@
...
+            // Arrange
+            Uri returnUrl = null;
+            using var controller = new AccountController(this._accountServiceMock.Object, this._loggerMock.Object);
+            controller.ControllerContext = new ControllerContext()
+            {
+                HttpContext = new DefaultHttpContext() { RequestServices = GetRequestService() }
+            };
+
+            // Act
+            var result = controller.Login(returnUrl);
```
Comment (Mejtika) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs:52]: `Login` action in the controller is awaitable, so in the current solution we're not waiting for it to finish and the test will pass with the current assertion because it's basically true. To get the real `result` we should consider awaiting this action. Currently, we're mocking the `RequestServices` and by doing that we need to register all the services that will be used in that controller action. The problem with awaiting that action is that we'll get the exception `No service for type 'Microsoft.AspNetCore.Mvc.ViewFeatures.ITempDataDictionaryFactory' has been registered.`, so we need to register also that type.
So basically we should consider awaiting all the awaitable actions that we're testing, but if we're also mocking `RequestServices` we need to register all the services the particular test needs. That is the price of unit testing controllers :(

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs [hunk 8]
```diff
@@ -0,0 +1,743 @@
...
+            using var controller = new AccountController(this._accountServiceMock.Object, this._loggerMock.Object);
+            var mockUrlHelper = new Mock<IUrlHelper>();
+            mockUrlHelper.Setup(x => x.Action(It.IsAny<UrlActionContext>())).Returns("callbackUrl");
+            controller.Url = mockUrlHelper.Object;
+
+            // Act
+            var result = controller.Login(model, returnUrl);
+
+            // Assert
+            var viewResult = Assert.IsType<RedirectToActionResult>(result.Result);
```
Comment (Mejtika) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs:132]: As above, we should consider awaiting that action and not waiting synchronously for its completion.

**Comment by szrtGreg** [comment]
@Mejtika  Thanks for helping me out. @WTobor tests are ready for merge.

**Review by WTobor — APPROVED** [review]
